proposals/theme_manifest.md

```diff
@@ -0,0 +1,60 @@
+# Theme manifest standards proposal
+
+## Description Of Issue
+Theme manifests should be documented and standardised. While themes are not part of Cumcord itself,
+they are an essential part of most users' client mod experience,
+so when theme loader plugins inevitably appear, the manifest (and repo) formats should be standardised and cross-compatible.
+
+## Solution
+Standardise and document the manifest format (new-BD-style), what the fields must be, and what each is for.
+
+## Advantages
+
+ - prevent 3rd parties from adding their own fields to manifests for their own use, cluttering repos.
+ - act as documentation for theme developers on what to include in their manifest
+
+## Disadvantages
+ - locks down possibly helpful info that can be included in manifests
+
+## General Implementation Details
+ - only new BD style manifests should be carried forward into CC themes - not `//META`-style manifests
+
+I propose the following TypeScript declaration, not because it has any direct use,
+but purely to concretely define the data structure required:
+```ts
+declare class ThemeManifest {
+ // The name of the theme
+ name:        string;
+ // A more detail description for extra info about the theme
+ description: string;
+ // The Discord ID of the theme's author - only one allowed, prioritised over author
+ authorId?:   string;
+ // Who wrote the theme; not required to be just one person
+ author:      string;
```

Comment:
Alright, regarding this, I would like to leave my thoughts.

I would prefer to drop the whole "CSS comment as a manifest" idea, and replace it with a normal manifest JSON file.

My main reason for this is making the manifest easily expandable whenever needed. (let us hope this is not a requirement all too often, though)

For example, assuming this interface is present:
```ts
interface Author {
  name: string;
  id: string;
}
```
I would give you the following suggestion:
```suggestion
 author: Author | Author[]
```

This can only be done if we drop the idea of using a CSS comment for the manifest, and instead, move to a more sensible solution, that being a JSON manifest.

Reply:
Note that, aside from this (rather big) suggestion, the rest of the manifest looks fine to me.

Reply:
i would assume said json manifest should still reside inside a comment?
```css
/* { 


} */
```

Reply:
I was actually aiming for a separate file, but I would need @Cr3atable's input on this before saying anything that is not agreed upon by all maintainers.

Reply:
We could do a similar file, similarly to how we do plugin loading (fetching plugin.json plugin.js, or in this case cumcord-theme.json, and theme.css) but I do also think that the comment format is easy enough to parse that it isn't worth removing. I think we'd have to talk to theme developers about what they'd prefer before continuing this proposal.

Reply:
Very well then.

Reply:
from ~~the perspective of~~ *my perspective as* a theme dev, i have no preference as i have had both formats previously (having built for bd & pc)

from ~~the perspective of~~ *my perspective as* a themer plugin implementer, arguably the separate file is actually easier to parse, but its not like either are difficult.

Reply:
The separate file implementation seems to make more sense from the perspective of Powercord's theme loader where themes are downloaded in whole as folders, but that's not what this would be. That's why I want to consider more.

Reply:
yes yes of course, though it does keep consistency with the current plugin loading system, where plugins are required to be hosted at `<base url>/plugin.js` with a manifest at `<base url>/plugin.json`

Reply:
to attempt to get the ball rolling on this again, I still think it makes the most sense from an implementation and flexibility perspective, as well as for consistency, to emulate the plugin system with, say, `theme.css` and `theme.json`

Reply:
My suggestion is that theme.css can be named anything arbitrarily, however the manifest must be at the same path and named `cumcord_theme.json`.

Example is as follows:
- https://example.com/path/themegoeshere.css
- https://example.com/path/cumcord_theme.json

This is to allow flexibility in css file names. The manifest could potentially specify the path of the css file, or it could be assumed by domain and path. Thoughts?

Reply:
This makes sense to me. It makes loading themes from a raw github repo much easier - with a repo with the following structure
```
- ROOT
  |- mytheme.theme.css (for bd compatibility)
  |- manifest.json (for pc/vz compatibility)
  |- cumcord_theme.json (for cc)
```
you can support all mods at once. The relevant paths would then become
 - `https://raw.githubusercontent.com/i_am_cool/mytheme/master/mytheme.theme.css`
 - `https://raw.githubusercontent.com/i_am_cool/mytheme/master/cumcord_theme.json`
Hence making the import url `https://raw.githubusercontent.com/i_am_cool/mytheme/master`

Reply:
Most themes will be hosted on a `github.io` subdomain because `@imports` require the correct mime type (text/css).

Reply:
oh true true

thinking about it i tend to use jsdelivr for gh imports for that reason

Reply:
I guess my main thing is this:
What would the import URL be?

If the import URL is the css file then the manifest can be inferred to be at `/cumcord_theme.json` on the same path.

If the import URL is the json file then the CSS file can be inferred from the path specified in the manifest.

If the import URL is the parent of both then the above inference from the manifest applies, which is somewhat odd, but consistent with plugins in that the path does not directly point towards the files.

Reply:
To follow up my last comment, the other benefit of the final solution is that the parent directory being the path allows you to serve HTML from that page, allowing for extensibility in regards to what gets displayed when people go to your plugin's link.

Reply:
i would argue that the manifest should contain a `path` tag pointing to the manifest, then the import url should be the manifest url

- inferring the manifest is too vague for my liking - id prefer things to be concrete
- using the parent url is consistent with plugins, but kind of an odd choice here as with plugins both files required had solid names, whereas here only the manifest does.

Reply:
however the point about serving a page does almost kind of win me over to the parent url idea on this one.

plus its cleaner to implement

Reply:
I vouch for making the manifest the entrypoint.

Reply:
I am bored so I wrote up some pros and cons of both solutions
| | Pros | Cons |
|-|-|-|
| Parent URL | a webpage can be displayed for clueless users, it is consistent with plugin urls | on the other hand, this may cause confusion as to if a url is a plugin or theme |
| Manifest URL | unambiguous, differentiates from plugins, clearly cumcord-specific | still not actually theme-specific, may confuse users as to why theme & plugins are different |

Reply:
The entrypoint should be the CSS file. No path tag should make it into the manifest.

Reply:
Sounds sensible, except if we end up with some kind of bundling system for themes in which case it may be necessary

Reply:
Weeelll we can always just copy the manifest into a dist directory like we do with plugins.

Reply:
Ig

Reply:
resolving this convo now.